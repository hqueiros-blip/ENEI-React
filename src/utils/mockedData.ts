// Mock dos eventos
export const events = [
  {
    id: 1,
    author: "Hugo Queirós",
    description: "React Basics & The Importance of Testing",
    scheduleTime: "9:30 às 11:30",
    image: "src/assets/reactPresentation.jpeg",
    attended: false,
  },
  {
    id: 2,
    author: "Mafalda Landeiro",
    description:
      "Kotlin KickStart: Mastering the Basics for Modern Development",
    scheduleTime: "9:30 às 11:30",
    image: "src/assets/kotlinPresentation.jpeg",
    attended: false,
  },
];
