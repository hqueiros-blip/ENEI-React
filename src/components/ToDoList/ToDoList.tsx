import { ToDo } from "types/toDo";
import { events } from "utils/mockedData";
import { ToDoItem } from "./ToDoItem";
import { useEffect, useState } from "react";
import { LoadingPage } from "layout/LoadingPage";

export const ToDoList = () => {
  const [todos, setTodos] = useState<ToDo[]>(events);
  const [isLoading, setIsLoading] = useState(true);

  const handleMarkAsAttended = (id: number) => {
    setTodos((prevTodos) =>
      prevTodos.map((todo) =>
        todo.id === id ? { ...todo, attended: true } : todo
      )
    );
  };

  useEffect(() => {
    setTimeout(() => {
      setIsLoading(false);
    }, 1000);
  }, []);

  if (isLoading) {
    return <LoadingPage />;
  }
  return (
    <div>
      <h1>Listagem de Apresentações</h1>

      {/* Mapeamento dos Eventos */}
      <ul>
        {todos.map((event) => (
          <ToDoItem
            eventEnei={event}
            key={event.id}
            onAttend={handleMarkAsAttended}
          />
        ))}
      </ul>
    </div>
  );
};
