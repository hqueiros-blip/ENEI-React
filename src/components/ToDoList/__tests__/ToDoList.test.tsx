import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { events } from "utils/mockedData";
import { ToDoList } from "../ToDoList";

describe("ToDoList Component", () => {
  it("Snapshot of List of events", async () => {
    const { asFragment } = render(<ToDoList />);

    await waitForElementToBeRemoved(screen.getByText(/Loading/i), {
      timeout: 3000,
    });
    expect(asFragment()).toMatchSnapshot();
  });
  it("Render List of events", async () => {
    render(<ToDoList />);

    await waitForElementToBeRemoved(screen.getByText(/Loading/i), {
      timeout: 3000,
    });
    // Usando os mocks verificar se todos os eventos estão visíveis
    events.forEach((event) => {
      expect(screen.getByText(event.description)).toBeInTheDocument();
    });
  });

  it("Mark a event as attended", async () => {
    const user = userEvent.setup();
    render(<ToDoList />);

    await waitForElementToBeRemoved(screen.getByText(/Loading/i), {
      timeout: 3000,
    });

    // Verificar se existe o botão para marcar o evento como assistido
    const button = screen.getAllByRole("button", {
      name: /Marcar evento como assistido/i,
    });

    // Uma vez que temos 2 eventos devemos ter 2 botões disponíveis
    expect(button).toHaveLength(2);

    // Clicar no 1º botão existente
    await user.click(button[0]);

    // Verificar que teve o comportamento esperado de o evento ser marcado como assistido
    await waitFor(() => {
      expect(screen.getByText("Evento Assistido ✅")).toBeInTheDocument();
    });
  });
});
