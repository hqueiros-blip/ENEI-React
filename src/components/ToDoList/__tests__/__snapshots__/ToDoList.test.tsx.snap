// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToDoList Component > Snapshot of List of events 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Listagem de Apresentações
    </h1>
    <ul>
      <li
        class="event__list"
      >
        <img
          alt="Hugo Queirós"
          class="event__image"
          src="/assets/reactPresentation.jpeg"
        />
        <div
          class="event__details"
        >
          <div>
            <h2>
              React Basics & The Importance of Testing
            </h2>
          </div>
          <div>
            <h3>
              Hugo Queirós
            </h3>
          </div>
          <div>
            <h4>
              Horário: 9:30 às 11:30
            </h4>
          </div>
          <button
            class="event__button"
          >
            Marcar evento como assistido
          </button>
        </div>
      </li>
      <li
        class="event__list"
      >
        <img
          alt="Mafalda Landeiro"
          class="event__image"
          src="/assets/kotlinPresentation.jpeg"
        />
        <div
          class="event__details"
        >
          <div>
            <h2>
              Kotlin KickStart: Mastering the Basics for Modern Development
            </h2>
          </div>
          <div>
            <h3>
              Mafalda Landeiro
            </h3>
          </div>
          <div>
            <h4>
              Horário: 9:30 às 11:30
            </h4>
          </div>
          <button
            class="event__button"
          >
            Marcar evento como assistido
          </button>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
